File: solutions/0020-valid-parentheses/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        for(int i=0;i<s.size();i++){
            if(s[i]=='(' or s[i]=='{' or s[i]=='['){
                stk.push(s[i]);
            }
            else if(s[i]==')'){
                if(stk.empty() or stk.top()!='(') return false;
                stk.pop();
            }
            else if(s[i]==']'){
                if(stk.empty() or stk.top()!='[') return false;
                stk.pop();
            }
            else if(s[i]=='}'){
                if(stk.empty() or stk.top()!='{') return false;
                stk.pop();
            }
        }
        if(!stk.empty()) return false;
        return true;
    }
};

File: solutions/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        mapper = {")":"(","}":"{","]":"["}
        stk = []
        for c in s:
            if c not in mapper:
                stk.append(c)
            else:
                if stk and stk[-1]==mapper[c]:
                    stk.pop()
                else:
                    return False
        if len(stk)==0:
            return True
        return False



        
        
